daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExceptionTest.scala

```diff
@@ -0,0 +1,536 @@
+         exception MyException1 = {
+           message \(e: M:MyException1) -> M:MyException1 {message} e
+         };
+
+        val e1 : AnyException = to_any_exception @M:MyException1 (M:MyException1 { message = "Message1" });
+        val t1 : Text = ANY_EXCEPTION_MESSAGE M:e1;
+
+         record @serializable MyException2 = { front: Text, back: Text } ;
+
+         exception MyException2 = {
+           message \(e: M:MyException2) -> APPEND_TEXT (M:MyException2 {front} e) (M:MyException2 {back} e)
+         };
+
+        val e2 : AnyException = to_any_exception @M:MyException2 (M:MyException2 { front = "Mess", back = "age2" });
+        val t2 : Text = ANY_EXCEPTION_MESSAGE M:e2;
+
+        val e3 : AnyException = to_any_exception @GeneralError (MAKE_GENERAL_ERROR "MessageG");
+        val t3 : Text = ANY_EXCEPTION_MESSAGE M:e3;
+
+        val e4 : AnyException = to_any_exception @ArithmeticError (MAKE_ARITHMETIC_ERROR "MessageA");
+        val t4 : Text = ANY_EXCEPTION_MESSAGE M:e4;
+
+        val e5 : AnyException = to_any_exception @ContractError (MAKE_CONTRACT_ERROR "MessageC");
+        val t5 : Text = ANY_EXCEPTION_MESSAGE M:e5;
+       }
+      """)
+
+    val testCases = Table[String, String](
+      ("expression", "expected-string"),
+      ("M:t1", "Message1"),
+      ("M:t2", "Message2"),
+      ("M:t3", "MessageG"),
+      ("M:t4", "MessageA"),
+      ("M:t5", "MessageC"),
+    )
+
+    forEvery(testCases) { (exp: String, res: String) =>
+      s"""eval[$exp] --> "$res"""" in {
+        val expected: SResult = SResultFinalValue(SValue.SText(res))
+        runExpr(pkgs)(e"$exp") shouldBe expected
+      }
+    }
+  }
+
+  "unhandled throw" should {
+
+    val pkgs: PureCompiledPackages = typeAndCompile(p"""
+       module M {
+
+         val unhandled1 : Update Int64 =
+            upure @Int64 (ADD_INT64 10 (throw @Int64 @GeneralError (MAKE_GENERAL_ERROR "oops1")));
+
+         val unhandled2 : Update Int64 =
+            try @Int64 (upure @Int64 (ADD_INT64 10 (throw @Int64 @GeneralError (MAKE_GENERAL_ERROR "oops2"))))
+            catch e -> None @(Update Int64);
+
+         record @serializable E1 = { } ;
+         exception E1 = { message \(e: M:E1) -> "E1" };
+         val unhandled3 : Update Int64 = upure @Int64 (throw @Int64 @M:E1 (M:E1 {})) ;
+
+         record @serializable E2 = { } ;
+         exception E2 = { message \(e: M:E2) -> throw @Text @M:E1 (M:E1 {}) } ; //throw from the message function
+         val unhandled4 : Update Int64 = upure @Int64 (throw @Int64 @M:E2 (M:E2 {})) ;
+
+       }
+      """)
+
+    val testCases = Table[String, SResult](
+      ("expression", "expected"),
+      ("M:unhandled1", SResultError(DamlEUnhandledException("oops1"))),
+      ("M:unhandled2", SResultError(DamlEUnhandledException("oops2"))),
+      ("M:unhandled3", SResultError(DamlEUnhandledException("E1"))),
+      ("M:unhandled4", SResultError(DamlEUnhandledException("E1"))),
```

Comment:
ah perfect, that’s exactly what I feared. We change the type of unhandled exceptions but not of handled exceptions if `message` crashes.

Reply:
I'm not really sure what you mean by _change the type_ here.
Is there anything to for now?

Reply:
If I add a `catch` at the top-level I get the actual exception, if I don’t catch it, I get the one from `DamlEUnhandledException` which comes from the `message` function and can have a different type. I’m fine with ignoring this for now but we should add TODOs/issues and address this (perhaps in combination with exposing the type and value in `DamlEUhandledException` directly).

Reply:
Yes. There is a TODO added.